lib/Runtime/Library/JavascriptGenerator.cpp

```diff
@@ -5,857 +5,509 @@
+            {
+                extractor->MarkVisitVar(curr);
+            }
         }
-        Var error = generator->CreateTypeError(JSERR_NonObjectFromIterable, scriptContext, _u("yield*"));
-        JavascriptOperators::OP_AsyncYield(generator, error, scriptContext);
-        return scriptContext->GetLibrary()->GetUndefined();
     }
 
-    Var JavascriptGenerator::AsyncGeneratorEnqueue(Var thisValue, ScriptContext* scriptContext, Var input, JavascriptExceptionObject* exceptionObj, const char16* apiNameForErrorMessage)
+    // args.Values is null when generator has been completed
+    if (this->args.Values != nullptr)
     {
-        // 1. Assert: completion is a Completion Record.
-        // 2. Let promiseCapability be ! NewPromiseCapability(%Promise%).
-        JavascriptPromise* promise = JavascriptPromise::CreateEnginePromise(scriptContext);
-
-        // 3. If Type(generator) is not Object, or if generator does not have an [[AsyncGeneratorState]] internal slot, then
-        //    a. Let badGeneratorError be a newly created TypeError object.
-        //    b. Perform ! Call(promiseCapability.[[Reject]], undefined, << badGeneratorError >>).
-        //    c. Return promiseCapability.[[Promise]].
-        if (!VarIs<JavascriptGenerator>(thisValue))
+        // mark argument variables for traversal
+        uint32 argCount = this->args.GetArgCountWithExtraArgs();
+        for (uint32 i = 0; i < argCount; i++)
         {
-            Var error = CreateTypeError(JSERR_NeedObjectOfType, scriptContext, apiNameForErrorMessage, _u("AsyncGenerator"));
-            promise->Reject(error, scriptContext);
-            return promise;
+            Js::Var curr = this->args[i];
+            if (curr != nullptr)
+            {
+                extractor->MarkVisitVar(curr);
+            }
         }
+    }
+}
 
-        JavascriptGenerator* generator = UnsafeVarTo<JavascriptGenerator>(thisValue);
+TTD::NSSnapObjects::SnapObjectType JavascriptGenerator::GetSnapTag_TTD() const
+{
+    return TTD::NSSnapObjects::SnapObjectType::SnapGenerator;
+}
 
-        if (!generator->GetIsAsync())
-        {
-            Var error = CreateTypeError(JSERR_NeedObjectOfType, scriptContext, apiNameForErrorMessage, _u("AsyncGenerator"));
-            promise->Reject(error, scriptContext);
-            return promise;
-        }
+void JavascriptGenerator::ExtractSnapObjectDataInto(TTD::NSSnapObjects::SnapObject* objData, TTD::SlabAllocator& alloc)
+{
+    TTD::NSSnapObjects::SnapGeneratorInfo* gi = alloc.SlabAllocateStruct<TTD::NSSnapObjects::SnapGeneratorInfo>();
 
-        // 4. Let queue be generator.[[AsyncGeneratorQueue]].
-        // 5. Let request be AsyncGeneratorRequest { [[Completion]]: completion, [[Capability]]: promiseCapability }.
-        AsyncGeneratorRequest* request = RecyclerNew(scriptContext->GetRecycler(), AsyncGeneratorRequest, input, exceptionObj, promise);
+    // TODO: BUGBUG - figure out how to determine what the prototype was
```

Comment:
Is this ok? Is TTD broken with this?

Reply:
@boingoing I believe TDD has always been broken on this path. I remember seeing this same comment but didn't have enough context to understand it further.

Reply:
Yes. All of this, including the TTD stuff (which I've got my eye on for possible simplification) was just copied over.